Add Service Worker to cache assets & requests

chrisdwheatley opened the pull request:
Here's my attempt at adding a Service Worker.

I'm unsure of what command runs in App Engine to start the server but since I've had to add a `package.json` to pull in the `sw-precache` dependency I figured using `npm start` to both generate the Service Worker & start the server might work well. If not then happy to take a steer on and amend.

addyosmani commented:
Thanks, @chrisdwheatley! @housseindjirdeh could probably confirm the likely behavior of kicking everything off on the server and how well this might work with the recent CI work too.

housseindjirdeh commented:
Thanks for submitting this @chrisdwheatley!!!!!

So I've got a [PR](https://github.com/tastejs/hacker-news-pwas/pull/28) up that fixes up a few errors with the CI pipeline. In the `script` step, I kick off a build and the output `_site` directory gets deployed to App Engine. 

You can just have an npm command that generates the service worker (`precache` for example) and call that command right after `bundle jekyll build` in `.travis.yml`. You may have to make sure that the service worker is generated within `_site` so you might need a `root: '_site'` parameter in `sw-precache-config`.

addyosmani commented:
@housseindjirdeh Head's up that I've just changed our domain DNS settings over to using Cloudflare so we get an SSL cert. We'll need HTTPS support for the Service Worker parts of this PR.

chrisdwheatley commented:
Just pushed some config changes based on @housseindjirdeh's suggestions. The Service Worker is built first in the `./site` directory and then moved over `./site/_site` with `bundle exec jekyll build`. This makes it easier to still be able to use & test the SW locally if desired, something I've added a note on how to do to `CONTRIBUTING.MD`.

I added the file to [`./site/app.yaml`](https://github.com/tastejs/hacker-news-pwas/pull/27/files#diff-85e450276b1213a4ba5439bf1323c487R63) but I'm not 100% sure that is needed for a generated file.

I had to try a few different methods to get Travis to switch from version `0.10` of Node.js to a newer version so apologies for the commit noise.

chrisdwheatley commented:
> Total precache size is about 0 B for 0 resources.

Just realised this workflow doesn't quite work, investigating an alternative way now.

chrisdwheatley commented:
For now I have a workaround to run `bundle exec jekyll build` twice, simply because we need the app to be built to know which files to cache but then need to make sure the service worker is in the right place.

The suggestion you had about setting the root directory as `_site` would solve this @housseindjirdeh but then it'd be more difficult to use the SW locally.

addyosmani commented:
Fwiw, https://hnpwa.com/ is now enabled. Deploying redirects for HTTP traffic so that once this is merged we're always on HTTPS. I'll defer to @housseindjirdeh for the notes above.

housseindjirdeh commented on .travis.yml:
```diff
@@ -2,9 +2,13 @@ language: ruby
+script:
+  - nvm install 7
+  - npm install
+  - bundle exec jekyll build
```
@chrisdwheatley yeah I see why you have to make sure the service worker file gets moved to the output folder but work locally as well. But I think we can still have one build call here for simplicity.

```
script:
  - bundle exec jekyll build
  - npm run precache
```

For the precache npm command, we can make sure it deploys to _site specifically (take a look at my comment in `package.json`.)

housseindjirdeh commented on .travis.yml:
```diff
@@ -2,9 +2,13 @@ language: ruby
-script: 
+script:
+  - nvm install 7
+  - npm install
```
I think both lines 8 and 9 probably makes more sense in the `before_install` step (after the first two commands already there)

housseindjirdeh commented on CONTRIBUTING.md:
```diff
@@ -8,4 +8,8 @@ If your Hacker News implementation meets the [specifications](https://github.com
+
+## Testing Service Worker locally
+
+To test the Service Worker locally you need to run `npm run precache` before running `bundle exec jekyll serve`.
```
Will just need to update this if you decide to change things up in package.json like I mentioned :P

housseindjirdeh commented on site/_includes/service-worker-register.html:
```diff
@@ -0,0 +1,11 @@
+  if ('serviceWorker' in navigator) {
+    window.addEventListener('load', () => {
+      navigator.serviceWorker.register('/service-worker.js').then(registration => {
+        // Registration was successful
```
I think console.log statements might be more helpful 👍

housseindjirdeh commented on site/_includes/service-worker-register.html:
```diff
@@ -0,0 +1,11 @@
+      navigator.serviceWorker.register('/service-worker.js').then(registration => {
+        // Registration was successful
+      }, err => {
+        // Registration failed
```
Same here

housseindjirdeh commented on site/package.json:
```diff
@@ -0,0 +1,10 @@
+  "name": "hacker-news-pwas",
+  "description": "Hacker News readers as Progressive Web Apps",
+  "scripts": {
+    "precache": "sw-precache --config=sw-precache-config.js --verbose"
```
So here I think we can have a separate call that users can use for locally starting the site. For example:

```
  "scripts": {
    "precache": "sw-precache --config=sw-precache-config.js --root=_site --verbose",
    "start": "sw-precache --config=sw-precache-config.js && bundle exec jekyll serve"
  },
```
So npm run precache will make sure the service worker file gets generated in `_site` if done after `bundle exec jekyll build` and for a user to just run the site locally with the service worker they can just run `npm start`. Let me know if you agree and think this is a better approach.

chrisdwheatley commented:
Thanks for all the great feedback @housseindjirdeh. Really appreciated another set of eyes on this as I was struggling to think of the best approach, your suggestions make things much cleaner.

The only bit which I had wrong initially was suggesting that running `npm run precache` locally before `bundle exec jekyll serve` would mean you could test the service worker, I don't think that'd have worked due to the code still needing to be built before running `sw-precache`. I've created an `npm run dev-sw` task to handle service worker generation locally but it feels like there still might be a nicer way to achieve what is required.

I've placed the `bundle exec` commands behind `npm run` scripts to keep all the commands in the same place, I hope that's ok.

housseindjirdeh commented on site/package.json:
```diff
@@ -0,0 +1,13 @@
+  "scripts": {
+    "build": "bundle exec jekyll build",
+    "dev": "bundle exec jekyll serve",
+    "dev-sw": "npm run build && sw-precache --config=sw-precache-config.js && npm run dev",
```
Don't think we need to run a build first. `sw-precache --config=sw-precache-config.js` should create the service worker file in the root of the project and `jekyll serve` should build the output folder along with the file.

`sw-precache --config=sw-precache-config.js && npm run dev` should be just fine (but please test it and let me know if that's not the case)

chrisdwheatley replied:
If we don't build first then the html & css hasn't been compiled or the assets moved. I tested by deleting the `_site` directory and running `sw-precache --config=sw-precache-config.js && npm run dev` which resulted in a service worker being generated but it not actually caching anything if that makes sense.

housseindjirdeh replied:
Oooooooooh yes you're right. Thanks for testing it thoroughly that makes perfect sense actually.

housseindjirdeh commented on site/package.json:
```diff
@@ -0,0 +1,13 @@
+  "description": "Hacker News readers as Progressive Web Apps",
+  "scripts": {
+    "build": "bundle exec jekyll build",
+    "dev": "bundle exec jekyll serve",
```
Do you think we need a command to run the site without a service worker? 🤔  I would think anytime a user wants to boot the site up - there's no reason to not include the service worker. I'm okay with removing this separation and having a single `dev` command thats generates the service worker and boots the site up.

Thinking out loud here however, I'm perfectly okay with leaving it as is if you think it's better :)

chrisdwheatley replied:
I don't mind a single command but I imagine it could be confusing if someone were to make some changes and not see them appear because the cached version of the file is being used (totally done this before plenty of times 🤕).

But then again it's good to have your dev env as close to prod as possible and 'Bypass for network' is a good way around the cache whilst still using the service worker.

housseindjirdeh replied:
Fair enough, I'm okay with having separate script commands as well 👍

housseindjirdeh commented:
LGTM :shipit: 

@addyosmani feel free to do a final pass through before we merge

addyosmani commented:
Thanks for working through the last pieces needed to get this across the finishing line folks <3

chrisdwheatley commented:
Would be good to get `master` deployed  now if that's possible.

housseindjirdeh commented:
@chrisdwheatley CI should be deploying with every merge :)

I restarted [your build](https://travis-ci.org/tastejs/hacker-news-pwas/builds/233539578) however in case you weren't seeing your changes. Feel free to always to take a look at the build status on Travis CI to make sure deployments went through.

chrisdwheatley commented:
My bad, one small issue in the CI flow which I've just pushed a fix up for now in #34.
